File: 506_Relative_Ranks.java
/*
https://leetcode.com/problems/relative-ranks/
Given scores of N athletes, find their relative ranks and the people with the top three highest scores, who will be
awarded medals: "Gold Medal", "Silver Medal" and "Bronze Medal".

Copy nums and sort, then create a map from each num to its rank. For each num in original order, lookup rank in map.
Convert rank to medal.
Time - O(n log n)
Space - O(n)
*/

class Solution {
    public String[] findRelativeRanks(int[] nums) {

        int n = nums.length;
        String[] ranks = new String[n];
        Map<Integer, String> medals = new HashMap<Integer, String>();
        medals.put(1, "Gold Medal");
        medals.put(2, "Silver Medal");
        medals.put(3, "Bronze Medal");

        int[] numsCopy = nums.clone();
        Arrays.sort(numsCopy);
        Map<Integer, Integer> numToRank = new HashMap<Integer, Integer>();

        for (int i = 0; i < n; ++i)
            numToRank.put(numsCopy[i], n - i);

        int rank;
        for (int i = 0; i < n; ++i) {
            rank = numToRank.get(nums[i]);
            ranks[i] = medals.getOrDefault(rank, Integer.toString(rank));
        }

        return ranks;
    }
}

File: python_1_to_1000/506_Relative_Ranks.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/relative-ranks/
# Given scores of N athletes, find their relative ranks and the people with the top three highest scores, who will be
# awarded medals: "Gold Medal", "Silver Medal" and "Bronze Medal".

# Create a list of tuples of num and its index in nums. Sort descending by num. Iterate over list, setting result
# with medals for the first 3 then rank as a string for others.
# Time - O(n log n)
# Space - O(n)

class Solution(object):
    def findRelativeRanks(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        num_i = [(num, i) for i, num in enumerate(nums)]
        num_i.sort(reverse=True)

        result = [None for _ in range(len(nums))]
        medals = ["Gold", "Silver", "Bronze"]

        for rank, (_, i) in enumerate(num_i):
            if rank < 3:                                # best ranked 3 get medals
                result[i] = medals[rank] + " Medal"
            else:
                result[i] = str(rank + 1)               # add 1 since enumeration is from zero

        return result
